app/models/user.rb

```diff
@@ -162,6 +164,45 @@ def remember_me?
     remember_token_expires_at && remember_token_expires_at > Time.zone.now
   end
 
+  def mfa_enabled?
+    !no_mfa?
+  end
+
+  def disable_mfa!
+    no_mfa!
+    self.mfa_seed = ''
+    self.mfa_recovery_codes = []
+    self.last_otp_at = nil
+    save!(validate: false)
+  end
+
+  def enable_mfa!(seed, level)
+    self.mfa_level = level
+    self.mfa_seed = seed
+    self.mfa_recovery_codes = Array.new(10).map { SecureRandom.hex(6) }
+    save!(validate: false)
+  end
+
+  def otp_verified?(otp)
+    if no_mfa?
+      true
+    elsif mfa_recovery_codes.include?(otp)
+      mfa_recovery_codes.delete(otp)
+      save!(validate: false)
+      true
+    else
+      totp = ROTP::TOTP.new(mfa_seed)
+      last_verification = totp.verify_with_drift_and_prior(otp, 30, last_otp_at)
+      if last_verification
+        self.last_otp_at = Time.at(last_verification).utc.to_datetime
+        save!(validate: false)
+        true
+      else
+        false
+      end
```

Comment:
Please move the code `else` to a new method. Also, we would want to do `mfa_recovery_codes.include?(otp)` at last, ie this logic should be in else and `totp.verify..` should be `elsif`.